### ISSUE TITLE
Snap based on dragged object in multi-selection

### ISSUE THREAD

**Issue by JoeCreates** [bjorn/tiled#2167]
Presently it seems like the top or left object in a selection of multiple objects is used for snapping. It would be better if the snapping were done based on the particular object in the selection was clicked for the drag.

**Comment by bjorn** [comment]
Right, it's not so much the top-left object, but the top-left of the bounding box of the selected objects.

Maybe you're right that it would be better to base it on the object clicked to start the drag. However, there isn't always such an object, since you can drag the current selection from anywhere while holding the Alt key.

**Comment by JoeCreates** [comment]
Ah I see. My workaround was to place a reference object far top left temporarily for snapping. I definitely think the clicked object makes far more sense, though. Top left of the selection is going to be so random.

Here's an idea for dragging from a non object: make it use the top left of the tile you clicked in? Well, this is equivalent to just maintaining the offset of the top left of the bounding box with whatever tile it lands in. This way your snapping will end up giving you an integer offset from the original position.

**Comment by bjorn** [comment]
> Here's an idea for dragging from a non object: make it use the top left of the tile you clicked in (or perhaps better put, it maintains the offset of the AABB inside it's tile, snapping to the same offset in whatever tile it lands in)? This way your snapping will end up giving you an integer offset (in tiles) from the original position.

I think snapping is expected to make sure things a grid-aligned. Making it move something by tile-distances doesn't sound like a good idea, or at least it could be quite annoying for people who're trying to get things snapped to the grid.

Maybe the behavior when doing Alt+Drag should just remain the same as it is now, and we'd only change the behavior when you drag by clicking a particular object.

**Comment by JoeCreates** [comment]
I have to disagree. I'm pretty sure nobody is really ever wanting to snap
the selection AABB to the grid. This is quite unintuitive, and I'm
struggling to think of an actual use for it. The AABB of a users  selection
isn't ever something they ever design. It is an unintended consequence of
the selection itself.

What is useful and expected is to be snapping objects to the grid. What my
suggestion does is it keeps all objects that were snapped to the grid
snapped. It is a way of preventing you from losing the offset that you have
tried to create.

If this happens to not be what you desire, but rather you want to snap a
specific object to the grid, you can also do that by dragging that object
specifically.

Inagine a scenario where you select objects A aligned on X to tile coords
(1, 1.6), and B at (2.3, 2), aligned to Y. If you leave the AABB snapping
as it is now then you end up with no way of maintaining your snapping at
all! You cannot drag an individual object because each would unsnap a
component of the other.

The only way to maintain A at x=1,2,3 etc and B at y=2,3,4 would be
something like my suggestion of mainting the offsets with respect to the
tile grid for the entire selection. This is surely more consistent with
what a user would expect.

**Comment by bjorn** [comment]
Alright, so let's extend the behavior of the Alt modifier to not only force a drag operation, but also to change the snapping behavior to snap based on distance moved rather than the position. When Alt is not pressed, there is always a particular object we can use for snapping.

I noticed also, that Inkscape snaps different corners of the object based on which one was the closest to the location at which you grabbed an object. And that even works with rotated objects. Of course, implementing that is going really far...

**Comment by JoeCreates** [comment]
Ah yeah. I think I noticed that when I saw the some object that was smaller
than a tile was snapping to the bottom left of tiles and another was
snapping to top left. I was trying to get the former to snap to top left.

I've realised that this was probably because one was a tile object and one
was a shape.

Snapping to the nearest edge for x and y separately would certainly be
useful to make this feel consistent. It's also useful if you have an object
smaller than a tile which you want sat perfectly on the ground in a
platformer, for example.

**Comment by bjorn** [comment]
> It actually snaps to sides based on your original click position with respect to the object bounds? That probably is going too far. xD Snapping to the nearest tile corner instead of always to a specific corner would still be useful though I think.

Well, they would probably take around the same development effort. I think snapping based on where you clicked the object might feel a little bit more stable, since it's more deterministic. Especially for polygons, there may be a lot of points that could snap to the grid, and snapping based on the one that is nearest to a grid position may not be the most helpful thing to do.

**Comment by JoeCreates** [comment]
Ah, for a polygon I might have expected it to be the bounding box in the
case that you simply have the shape selected. If you have points selected
on the other hand then probably the point nearest where you clicked, which
is a special case requiring the use of a different tool to the normal
object selection.

Snapping to the nearest edges with just the bounding box probably wouldnt
be too complicated most of the time. An exception which may br fiddly is if
you have an object which is only just smallerbor larger than a tile. This
would make the threshold to snap to a particular side very small, so that
would certainly be an argument for click-position based snapping. I would
suggest maybe the cursor icon change to indicate which sides are going to
be snapping.

Another use case to consider with snapping is that you may wish to snap
only on one axis and not lose the non-grid-aligned position the other axis.
I think a good solution to this might be to have another key modifier
(shift) to make the movement affect only the axis which has moved the most,
so you will always be moving only in a straight line vertically or
horizontally.
